src/compiler/crystal/command/format.cr

```diff
@@ -47,156 +34,157 @@ class Crystal::Command
-      exit(result == source ? 0 : 1) if check_files
 
-      print result
-    rescue ex : InvalidByteSequenceError
-      STDERR.print "Error: ".colorize.toggle(@color).red.bold
-      STDERR.print "source is not a valid Crystal source file: ".colorize.toggle(@color).bold
-      STDERR.puts ex.message
-      exit 1
-    rescue ex : Crystal::SyntaxException
-      if @format == "json"
-        STDERR.puts ex.to_json
-      else
-        STDERR.puts ex
+    private def format_many(files)
+      files.each do |filename|
+        format_file_or_directory filename
       end
-      exit 1
-    rescue ex
-      couldnt_format "STDIN", ex
-      STDERR.puts
-      exit 1
-    end
-  end
-
-  private def format_many(files, check_files, excludes)
-    files.each do |filename|
-      format_file_or_directory filename, check_files, excludes
     end
-  end
 
-  private def format_file_or_directory(filename, check_files, excludes)
-    if File.file?(filename)
-      unless excludes.any? { |exclude| filename.starts_with?(exclude) }
-        format_file filename, check_files
+    private def format_file_or_directory(filename)
+      if File.file?(filename)
+        unless @excludes.any? { |exclude| filename.starts_with?(exclude) }
+          format_file filename
+        end
+      elsif Dir.exists?(filename)
+        filename = filename.chomp('/')
+        filenames = Dir["#{filename}/**/*.cr"]
+        format_many filenames
+      else
+        error "file or directory does not exist: #{filename}"
       end
-    elsif Dir.exists?(filename)
-      filename = filename.chomp('/')
-      filenames = Dir["#{filename}/**/*.cr"]
-      format_many filenames, check_files, excludes
-    else
-      error "file or directory does not exist: #{filename}"
     end
-  end
 
-  private def format_file(filename, check_files)
-    source = File.read(filename)
+    private def format_file(filename)
+      source = File.read(filename)
+      format_source filename, source
+    end
 
-    begin
-      result = Crystal.format(source, filename: filename)
+    private def format_source(filename, source)
+      result = format(filename, source)
+      @stdout.print result if @format_stdin
       return if result == source
 
-      if check_files
-        check_files << FormatResult.new(filename, FormatResult::Code::FORMAT)
-      else
-        File.write(filename, result)
-        STDOUT << "Format".colorize(:green).toggle(@color) << ' ' << filename << '\n'
-      end
+      on_format filename, result
     rescue ex : InvalidByteSequenceError
-      if check_files
-        check_files << FormatResult.new(filename, FormatResult::Code::INVALID_BYTE_SEQUENCE)
-      else
-        STDERR.print "Error: ".colorize.toggle(@color).red.bold
-        STDERR.print "file '#{Crystal.relative_filename(filename)}' is not a valid Crystal source file: ".colorize.toggle(@color).bold
-        STDERR.puts ex.message
-      end
+      on_invalid_byte_sequence_error filename, ex
```

Comment:
Why not inline these methods?

Reply:
It's for testability however it is not used for now.

I thought, adding such a class in spec file when I feel stdio test is hard.

```crystal
class CaptureFormatCommand
  getter captures = [] of {type: Symbol, filename: String, ex: Exception?}
  def on_invalid_byte_sequence_error(filename, ex)
    captures << {type: :invalid_byte_sequence_error, filename: filename, ex: ex}
  end

  # Override `on_format`, `on_syntax_error` and `on_bug`...
end
```

But stdio test is enough to check errors, so this approach is not used in fact.